plugins/Toolbox/src/CloudSync/DownloadPresenter.py

```diff
@@ -0,0 +1,127 @@
+import os
+import tempfile
+from functools import reduce
+from typing import Dict, List, Optional, Any
+
+from PyQt5.QtNetwork import QNetworkReply
+
+from UM import i18n_catalog
+from UM.Logger import Logger
+from UM.Message import Message
+from UM.Signal import Signal
+from UM.TaskManagement.HttpRequestManager import HttpRequestManager
+from cura.CuraApplication import CuraApplication
+from plugins.Toolbox.src.UltimakerCloudScope import UltimakerCloudScope
+from plugins.Toolbox.src.CloudSync.SubscribedPackagesModel import SubscribedPackagesModel
+
+
+## Downloads a set of packages from the Ultimaker Cloud Marketplace
+# use download() exactly once: should not be used for multiple sets of downloads since this class contains state
+class DownloadPresenter:
+
+    def __init__(self, app: CuraApplication):
+        # Emits (Dict[str, str], List[str]) # (success_items, error_items)
+        # Dict{success_package_id, temp_file_path}
+        # List[errored_package_id]
+        self.done = Signal()
+
+        self._app = app
+        self._scope = UltimakerCloudScope(app)
+
+        self._started = False
+        self._progress_message = self._createProgressMessage()
+        self._progress = {}  # type: Dict[str, Dict[str, Any]] # package_id, Dict
+        self._error = []  # type: List[str] # package_id
+
+    def download(self, model: SubscribedPackagesModel):
+        if self._started:
+            Logger.error("Download already started. Create a new %s instead", self.__class__.__name__)
+            return
+
+        manager = HttpRequestManager.getInstance()
+        for item in model.items:
+            package_id = item["package_id"]
+
+            request_data = manager.get(
+                item["download_url"],
+                callback = lambda reply, pid = package_id: self._onFinished(pid, reply),
+                download_progress_callback = lambda rx, rt, pid = package_id: self._onProgress(pid, rx, rt),
+                error_callback = lambda rx, rt, pid = package_id: self._onProgress(pid, rx, rt),
+                scope = self._scope)
+
+            self._progress[package_id] = {
+                "received": 0,
+                "total": 1,  # make sure this is not considered done yet. Also divByZero-safe
+                "file_written": None,
+                "request_data": request_data
+            }
+
+        self._started = True
```

Comment:
I'm not sure if this method will ever be called in such a way that it matters, but as written it's not thread-safe. (Or at least the defensive programming 'only run this once' isn't.)

Reply:
because there is no lock on executing this function? It was not designed to be thread-safe. You suggest something like this on the download function, I guess?

https://stackoverflow.com/a/39147428

Reply:
> It was not designed to be thread-safe.

I was thinking more about future adaptations, but it's not a big issue if it's not threaded right now.

(I wish there was a good way to indicate the 'seriousness' of a review comment, from 'needs fixing immediately' to 'just making sure you're aware of this' the latter of which are the majority of my comments in this.)

Reply:
Also yes, the lock is a way to solve it.

Reply:
I think I'll leave it the way it is. I expect a singular SyncOrchestrator with a singular DownloadPresenter